### PR TITLE
Use chroot path for initrd and kernel configs, filenames for drives

### PR DESCRIPTION
**Pull request by loony-bean**
`kernel_image_path` and `initrd_path` from BootSource should be populated from chrooted file path when we build the boot source config. `path_on_host` field on Drive is also updated to carry only drive filename, but only right before sending object to the socket.

Fixes a regression from https://github.com/blockjoy/firec/commit/196c88b3e026a4db645e2182df053612130f6f55.

### CODE REVIEW

File: src/config/drive.rs [hunk 1]
```diff
@@ -57,6 +60,11 @@ impl<'d> Drive<'d> {
     }
 
     /// Host level path for the guest drive.
+    pub fn host_path(&self) -> &Path {
```
Comment (zeenix) [inline src/config/drive.rs:63]: `path_on_host` is supposed to be host path so I'm not sure why we need another field/getter for the same thing. Perhaps `path_on_guest` (or just `path`) is needed here?

File: src/config/mod.rs [hunk 2]
```diff
@@ -32,8 +35,10 @@ pub struct Config<'c> {
     metrics_path: Option<Cow<'c, Path>>,
     metrics_fifo: Option<Cow<'c, Path>>,
     pub(crate) jailer_workspace_dir: Cow<'c, Path>,
+    pub(crate) kernel_image_file: Cow<'c, Path>,
```
Comment (zeenix) [inline src/config/mod.rs:38]: `kernel_image_filename` would be better and consistent with `std` and `tokio` naming scheme.

File: src/config/mod.rs [hunk 3]
```diff
@@ -93,8 +100,8 @@ impl<'c> Config<'c> {
     /// Create boot source from `self`.
     pub fn boot_source(&self) -> BootSource<'_> {
         BootSource {
-            kernel_image_path: self.kernel_image_path.as_ref(),
-            initrd_path: self.initrd_path.as_ref().map(AsRef::as_ref),
+            kernel_image_path: self.kernel_image_file.as_ref(),
+            initrd_path: self.initrd_file.as_ref().map(AsRef::as_ref),
```
Comment (zeenix) [inline src/config/mod.rs:104]: These are expected to be paths (relative to the chroot), not just filenames so I'm a bit confused about the rationale for these changes here. My bad that I didn't notice in the review but please change it ASAP.

File: src/config/mod.rs [hunk 4]
```diff
@@ -76,8 +81,10 @@ impl<'c> Config<'c> {
             metrics_path: None,
             metrics_fifo: None,
             jailer_workspace_dir: Path::new("/srv/jailer/firecracker/root").into(),
+            kernel_image_file: Path::new(KERNEL_IMAGE_FILENAME).into(),
             kernel_image_path: kernel_image_path.into(),
             host_kernel_image_path: Path::new("/srv/jailer/firecracker/root/debian-vmlinux").into(),
```
Comment (zeenix) [inline src/config/mod.rs:86]: I see you've hardcoded the paths in 196c88b for some reason. That's not only bad but also breaks this API.

**Review by zeenix — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -21,7 +21,7 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         .exec_file(Path::new("/usr/bin/firecracker"))
         .build();
 
-    let root_drive = Drive::builder("root", Path::new("debian.ext4"))
+    let root_drive = Drive::builder("root", Path::new("/tmp/debian.ext4"), Path::new("debian.ext4"))
```
Comment (zeenix) [inline README.md:24]: user shouldn't need to specify the same paths twice. User should only specify the path and if we're dealing with jailer, we build the host path ourselves.

File: src/config/mod.rs [hunk 6]
```diff
@@ -91,12 +91,22 @@ impl<'c> Config<'c> {
...
+                Some(Path::new(initrd_filename))
+            }
+            None => None,
+        };
+
+        let kernel_image_file = Path::new(KERNEL_IMAGE_FILENAME);
+
+        Ok(BootSource {
+            kernel_image_path: kernel_image_file,
+            initrd_path: initrd_filename,
```
Comment (zeenix) [inline src/config/mod.rs:104]: if these needs to be filenames, then the fields need renaming.

File: src/config/drive.rs [hunk 7]
```diff
@@ -22,17 +30,17 @@ pub struct Drive<'d> {
...
-        P: Into<Cow<'d, Path>>,
     {
         DriveBuilder(Drive {
             drive_id: drive_id.into(),
             is_read_only: false,
             is_root_device: false,
             part_uuid: None,
-            path_on_host: path_on_host.into(),
+            path: Path::new("").into(),
+            path_on_host: Path::new("").into(),
```
Comment (zeenix) [inline src/config/drive.rs:43]: At least one of them is a must for user to provide, which is why it was a parameter to `builder`.

File: src/config/mod.rs (hunk 6 above)
Comment (loony-bean) [reply src/config/mod.rs:104]: The names are part of socket API

Comment (zeenix) [reply src/config/mod.rs:104]: ah right, i forgot again we're serializing the config struct. :facepalm: 

Let's make sure these need to be filenames.

File: README.md [hunk 8]
```diff
@@ -21,7 +21,8 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         .exec_file(Path::new("/usr/bin/firecracker"))
         .build();
 
-    let root_drive = Drive::builder("root", Path::new("debian.ext4"))
+    let root_drive = Drive::builder("root")
+        .path(Path::new("/tmp/debian.ext4"))?
```
[the remark this answers is not in the document]
Comment (loony-bean) [reply README.md]: fixed

File: src/config/drive.rs (hunk 7 above)
Comment (loony-bean) [reply src/config/drive.rs]: yeah fixed

File: src/config/drive.rs [hunk 9]
```diff
@@ -56,9 +69,9 @@ impl<'d> Drive<'d> {
         self.part_uuid.as_deref()
     }
 
-    /// Host level path for the guest drive.
-    pub fn path_on_host(&self) -> &Path {
-        &self.path_on_host
+    /// Absolute path for the guest drive.
```
Comment (zeenix) [inline src/config/drive.rs:72]: That's the same as "Host level path for the guest drive.", which I copied from upstream fc docs.

Comment (loony-bean) [reply src/config/drive.rs]: it's a different field now (`path`)
updated a comment a bit

File: src/machine.rs [hunk 10]
```diff
@@ -318,7 +318,16 @@ impl<'m> Machine<'m> {
...
             let url: hyper::Uri = Uri::new(&self.config.host_socket_path, &path).into();
-            let json = serde_json::to_string(&drive)?;
+
+            // Send modified drive object, with drive file in chroot location
+            let mut drive_obj = drive.clone();
+            let drive_filename = drive_obj
+                .path_on_host()
+                .file_name()
+                .ok_or(Error::InvalidDrivePath)?
+                .to_owned();
```
Comment (zeenix) [inline src/machine.rs:327]: do we really need it to be owned?

File: src/config/mod.rs [hunk 11]
```diff
@@ -393,25 +401,48 @@ mod tests {
             .build()
             .unwrap();
 
-        assert_eq!(config.initrd_path.unwrap().as_os_str(), "/tmp/initrd.img");
+        assert_eq!(
+            config.initrd_path.as_ref().unwrap().as_os_str(),
+            "/tmp/initrd.img"
+        );
```
Comment (zeenix) [inline src/config/mod.rs:410]: why is this change needed? :thinking:

File: src/config/mod.rs [hunk 12]
```diff
@@ -393,25 +401,48 @@ mod tests {
...
         assert_eq!(
             config
                 .host_initrd_path
+                .as_ref()
```
Comment (zeenix) [inline src/config/mod.rs]: and this?

File: src/config/mod.rs (hunk 11 above)
Comment (zeenix) [reply src/config/mod.rs:410]: Looking at the next commit, I realized you probably made a mistake with splitting. This change isn't needed in this commit (probably some others too).

File: src/config/mod.rs (hunk 12 above)
Comment (loony-bean) [reply src/config/mod.rs]: because of this line, where config is borrowed
https://github.com/blockjoy/firec/pull/14/commits/cbd16a43e282c82dd4e2da131a145fc87b593192#diff-b0ba3e54b6f5ef1fa6aed8075e352a68fda66067fbafa8421042a2347801338aR443

Comment (loony-bean) [reply src/config/mod.rs]: another way around is to call host_initrd_path() i suppose

File: src/config/mod.rs [hunk 13]
```diff
@@ -205,15 +205,14 @@ impl<'c> Config<'c> {
...
 #[derive(Debug, Serialize, Deserialize)]
-pub struct BootSource<'b> {
+pub struct BootSource {
     /// The kernel image path.
-    #[serde(borrow)]
-    pub kernel_image_path: &'b Path,
+    pub kernel_image_path: PathBuf,
     /// The (optional) kernel command line.
-    pub boot_args: Option<&'b str>,
+    pub boot_args: Option<String>,
```
Comment (zeenix) [inline src/config/mod.rs:212]: Why? we don't need to modify the boot_args.

File: src/config/mod.rs [hunk 14]
```diff
@@ -88,21 +88,21 @@ impl<'c> Config<'c> {
     }
```
Comment (zeenix) [inline src/config/mod.rs:89]: Some rationale would be good to have on " Put relative paths starting with / into BootSource objects" commit.

File: src/config/mod.rs (hunk 12 above)
Comment (zeenix) [reply src/config/mod.rs]: > because of this line, where config is borrowed
https://github.com/blockjoy/firec/commit/cbd16a43e282c82dd4e2da131a145fc87b593192#diff-b0ba3e54b6f5ef1fa6aed8075e352a68fda66067fbafa8421042a2347801338aR443

Right so just need to be moved into that commit.

Comment (zeenix) [reply src/config/mod.rs]: > another way around is to call host_initrd_path() i suppose

That'd be better I think.

Comment (loony-bean) [reply src/config/mod.rs]: oh we don't have that func :)

Comment (loony-bean) [reply src/config/mod.rs]: i mean in this commit, and in the following we remove those as_ref's

Comment (loony-bean) [reply src/config/mod.rs]: > Right so just need to be moved into that commit.

it's related to changes in this commit

Comment (zeenix) [reply src/config/mod.rs]: how so, the type of those fields is not changing and neither is the string you're comparing with here.

**Review by zeenix — APPROVED** [review]
